### ISSUE TITLE
sim.data does not contain chip <-> host connections

### ISSUE THREAD

**Issue by arvoelke** [nengo/nengo-loihi#154]
Discovered while trying to find a work-around to #152.

It may be useful to access the `BuiltConnection` object corresponding to connections from chip to host, for example to obtain the decoding weights. However, such an object cannot be found in `sim.data`, and neither can the connection from host to chip:

```python
import nengo
import nengo_loihi

with nengo.Network() as model:
    u = nengo.Node(output=0)
    x = nengo.Ensemble(100, 1)
    y = nengo.Node(size_in=1)
    
    conn1 = nengo.Connection(u, x)
    conn2 = nengo.Connection(x, y)
    
with nengo_loihi.Simulator(model) as sim:
    sim.run(1.0)    

print(conn1 in sim.data)
print(conn2 in sim.data)
```

```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-93-2bd3edf8322b> in <module>()
     13     sim.run(1.0)
     14 
---> 15 print(conn1 in sim.data)
     16 print(conn2 in sim.data)

~/anaconda3/envs/py36/lib/python3.6/_collections_abc.py in __contains__(self, key)
    664     def __contains__(self, key):
    665         try:
--> 666             self[key]
    667         except KeyError:
    668             return False

~/CTN/nengo-loihi/nengo_loihi/simulator.py in __getitem__(self, key)
     45                     target = fallback.raw
     46                     break
---> 47         assert key in target, "probed object not found"
     48 
     49         if (key not in self._cache

AssertionError: probed object not found
```

Relatedly, the above stack trace reveals another error: `__getitem__` is raising an `AssertionError`, instead of a `KeyError`, and so the `conn1 in sim.data` is raising an error instead of returning `False`. `conn2 in sim.data` also raises the same error for the same reason (not found).

Swapping the simulator for `nengo.Simulator` prints `True` twice, as expected.

**Comment by tbekolay** [comment]
This is a known issue (https://github.com/nengo/nengo-loihi/blob/master/nengo_loihi/tests/test_simulator.py#L45).

It happens because [chip -> host](https://github.com/nengo/nengo-loihi/blob/master/nengo_loihi/splitter.py#L337) and [host -> chip](https://github.com/nengo/nengo-loihi/blob/master/nengo_loihi/splitter.py#L232) connections no longer exist after the splitting process because those connections are replaced with at least two connections, one on the host side and one on the chip side, both of which connect to/from nodes that do the transfer of information.

Because of that, the fix is not straightforward. The fix, in my opinion, is to move the logic currently done in the splitter system into the builder. Others have argued essentially the opposite, to do as much as possible with frontend objects, in which case the fix to this issue is not clear to me... I suppose you could make some kind of connection proxy object that connects the original frontend connection to the signals that were created at some point in the builder?

Anyway, we're going to have to commit one way or another soon, as we accruing a fair bit of technical debt at this point.

**Comment by tcstewar** [comment]
> Others have argued essentially the opposite, to do as much as possible with frontend objects, in which case the fix to this issue is not clear to me..

As one of those "others", the way I've been picturing is to maintain a dictionary that maps from original_connections -> splitter connections, and then have the `sim.data[]` `__getitem__` method be smart enough to follow that mapping.  Things get a little tricky if we have to merge multiple `data` items together, but that's something that we'd have to do if we merged the splitter and builder in any case.

**Comment by tbekolay** [comment]
> maintain a dictionary that maps from original_connections -> splitter connections, and then have the `sim.data[]` `__getitem__` method be smart enough to follow that mapping

We already do this for objects, because there's a one-to-one mapping from objects in the original network and objects in the split networks. In the case of host->chip and chip->host connections, however, this doesn't work because each connection is split into multiple connections. So when you probe the weights on that connection, for example, it's not clear which of the two connections should end up being probed, so we would need custom logic there (which I was thinking would be encapsulated in the connection proxy object).

**Comment by tcstewar** [comment]
> So when you probe the weights on that connection, for example, it's not clear which of the two connections should end up being probed, so we would need custom logic there (which I was thinking would be encapsulated in the connection proxy object).

Yes -- that "which of the two connections" issue is what I was thinking about, but you said it much more clearly than I did.  :)  I was thinking of encapsulating that logic in the `__getitem__` method, but I could also see it happening in the proxy object.

**Comment by tbekolay** [comment]
Ah, gotcha 😄 I'm also in the back of my mind a bit worried that the thing you will want to probe will just straight up not exist post-splitting, but as things are now all you can probe is input, output, and weights, and those should always be available. So just paranoia on my part!

**Comment by arvoelke** [comment]
This issue is also overlapping with #4 (the latter points out the same for `sim.model.param`).
